### DISCUSSION TITLE
Dealing with line based languages

### DISCUSSION THREAD

**Discussion by katawful** [tree-sitter/tree-sitter#1424] [Q&A] [answered]
I'm picking up development on a parser I've been chipping away at for a few months now, and a big problem I've been having is dealing with end of line. For this language, each line must be complete. It can't carry to the next line. So something like the following will fail to compile by the compiler:

```
if (some condition
   that carries to new line)
   ...
```
**fails**
```
if (some condition on same line)
  ...
```
**succeeds**

There is no ambiguity for this in the language. All lines must complete.  For my parser however, I can't seem to find a clean way to achieve this. I've been putting optional and non-optional end of line characters where possible in the rules where possible, but I can't force end of line everywhere:

```
if (condition) <- end of line must be here, nothing else but condition
  some cool function <- ambiguity in end of line here
endif <- same end of line
```
The function in this if block has a lot of ambiguity to it. That function could be called by itself just fine, **or** it could be called within another call, say passing it to a variable or using it as an argument for another function. Thus I can't just call an end of line character at the end of this function in my parser rules. That leads me with imprecise grammar that leads to a ton of ambiguity when developing, and complexity on top of that and I have been unable to find a way to achieve this. I'm not aware of many languages that are line based. I know bash sorta is. It achieves this with a scanner, but is there a simpler manner within just the grammar file?

**Comment by ahelwer** [comment, marked as answer]
What happens if you make a `line: $ => seq($.line_expr, $.newline)` rule where `line_expr` is every possible expression on a single line?

**Reply by katawful** [reply]
It doesn't end the ambiguity, expressions and statements can still carry over to the next line if the parsing fits. I get more luck just putting newline rules in as many places as possible, but this is quite hard to keep track of

**Reply by ahelwer** [reply]
Did you remove `\n` from extras?

**Reply by katawful** [reply]
Ah, that was it. That seems to work, gotta work out more testing to make sure

**Reply by katawful** [reply]
Ok, testing seems to work out. Thanks again!

**Reply by ahelwer** [reply]
@katawful no problem! Remember to handle windows newlines with `/\r?\n/`.

**Reply by theHamsta** [reply]
Mac newlines have only \r

**Reply by ahelwer** [reply]
@theHamsta really???? oh my god that's horrifying, I had no idea.

**Reply by maxbrunsfeld** [reply]
Only very very old mac system used `\r` though. There's pretty much no need to support it today.
